### PR TITLE
Integration tester skeleton

### PR DESCRIPTION
**Pull request by aaomidi**
TODO:

- [x] Add better logging
- [x] Add CI/CD pipelines (doing this in another PR)
- [x] Readme

Integration tester skeleton will replace the facilitator code that we're running on terraform to generate samples. This code is responsible for finding all the various environment information by reading  a manifest file, and then generate the suitable samples for it.

#241 depends on this PR.

### CODE REVIEW

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [inline facilitator/src/bin/facilitator.rs:1034]: I can probably do this in another PR if we don't want to take it here

Comment (tgeoghegan) [reply facilitator/src/bin/facilitator.rs:1034]: No, this a much better error message, and is useful to have here. However I would just do `.context("failed to parse batch signing key")?`. There's no need for a closure since you're not capturing any variables here.

File: integration-tester/cmd/root.go
Comment (tgeoghegan) [inline integration-tester/cmd/root.go:68]: explorer? Like Windows explorer?

File: integration-tester/kubernetes/kubernetes.go
Comment (tgeoghegan) [inline integration-tester/kubernetes/kubernetes.go:1]: I added a Kubernetes module to `workflow-manager` in #292 that shares a little bit of functionality with this file, namely the bits that get a client. It'd be nice to share modules between different go targets.

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [inline integration-tester/tester/logic.go:36]: You'll need to introduce something that deletes these jobs when they're completed, or we'll run into the same problems we did with intake jobs where we have too many completed jobs lying around.

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [inline integration-tester/tester/logic.go:115]: I think go style frowns on abbreviations like this, so `getValidPacketDecryptionKey` would be better. Also what makes a key valid? From the implementation, it looks like that means whatever key is currently advertised in a manifest, but it'd be good to explain that in a doccomment.

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [inline integration-tester/tester/logic.go:132]: Similar comment to `getValidPDK`: expand the acronym and document the basis on which a key is deemed valid

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
Can you help me understand what the goal of this PR is? It looks like what's achieved is to build an image that can replace the `sample-maker` cronjobs and doesn't need to use fixed packet encryption keys. Is that accurate? If so, then do you want to include the Terraform changes to replace the `sample-makers` with this in this PR or in a subsequent one? Basically I'd love to have a sense of what your broader plan is so I know how this PR fits in with that vision.

File: integration-tester/cmd/root.go
Comment (aaomidi) [reply integration-tester/cmd/root.go:68]: At this point I have no idea what I meant by this comment 😅

File: integration-tester/kubernetes/kubernetes.go
Comment (aaomidi) [reply integration-tester/kubernetes/kubernetes.go:1]: I'll wait for that to get merged in and pull that module in here

File: integration-tester/tester/logic.go
Comment (aaomidi) [reply integration-tester/tester/logic.go:36]: Yep!

File: integration-tester/cmd/root.go
Comment (aaomidi) [reply integration-tester/cmd/root.go:68]: Ah this is just some code I share between projects. Irrelevant to this one.

**Comment by aaomidi** [comment]
> Is that accurate? If so, then do you want to include the Terraform changes to replace the sample-makers with this in this PR or in a subsequent one? Basically I'd love to have a sense of what your broader plan is so I know how this PR fits in with that vision.

Yep that's accurate.

I'm looking to make this PR not change anything in terraform yet. I'm planning to update terraform either in #241 or in its own PR (I'll probably make it in its own PR, #241 is a major change).

The work in #241 showed that an integration-tester is necessary since terraform is eventually not going to be responsible for making the various testing keys and we want to test with actual manifest files and the actual deploy-operator work.

File: integration-tester/kubernetes/kubernetes.go
Comment (aaomidi) [reply integration-tester/kubernetes/kubernetes.go:1]: Yep it's fine! Just when building in docker the entire context of the git repo needs to be sent to the build agent, rather than this project alone.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
My big concerns here are the keys that the ingestion batch maker will use. It should not be using the data share processor's batch signing key (ingestors have their own batch signing key), and it should be discovering the packet encryption keys via the data share processor specific manifests instead of peering into Kubernetes secrets or using hard coded keys.

File: integration-tester/tester/types.go
Comment (aaomidi) [reply integration-tester/tester/types.go:1]: `root.go` and `run.go` is just how cobra & viper recommend their project layouts. `main.go` just needs to exist to bootstrap the application.

I don't necessarily agree that putting everything in `main.go` makes it more straightforward.

Right now, for example, if we find out we would rather pass something as a config value, or environment variable into workflow-manager there would be a lot of refactoring that would need to happen. There's also dozens of lines of `log.FatalF` for error management that we can skip over with this layout. These extra steps just make it really easy to follow the structure of the application once familiarity with viper & cobra is established.

`root` being the command registry.
`run` being the one specific command.

I do agree with logic.go and types.go though. That's just personal preference of keeping basic type definitions outside of the core logic of the module. But I can consolidate them into a single file and put them under `run.go`.

File: integration-tester/tester/logic.go
Comment (aaomidi) [inline integration-tester/tester/logic.go:134]: Let's just double check to make sure this makes sense

File: integration-tester/tester/logic.go
Comment (aaomidi) [inline integration-tester/tester/logic.go:207]: I'll write a test for this piece

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [inline integration-tester/tester/logic.go:40]: nit: expand this to `batchSigningKey`

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [reply integration-tester/tester/logic.go:134]: Based on what's currently in `kubernetes.tf`, I think you have `own` and `peer` reversed. It's possible that `facilitator generate-ingestion-sample` is just wrong in its treatment of `own` and `peer` vs. facilitator and PHA.

I'm also not sure about the identity arguments, because I'm not sure where `facilManifest.IngestionIdentity` or `phaManifest.IngestionIdentity` come from, or what their values will be. But consider carefully whether they need values at all, and what their values should be based on whether the bucket being written to is a GCS or S3 bucket. The scheme for all these buckets and identities is documented [here](https://docs.google.com/document/d/1MdfM3QT63ISU70l63bwzTrxr93Z7Tv7EDjLfammzo6Q/edit#heading=h.m8nz1l5m6znk).

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [inline integration-tester/tester/logic.go:159]: Are these functions ever used outside this module? Do they need to be public?

File: integration-tester/tester/logic.go
Comment (tgeoghegan) [inline integration-tester/tester/logic.go:230]: Suggest renaming this type to `IngestorGlobal` so that then it reads as `manifest.IngestorGlobal`.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: facilitator/src/sample.rs
Comment (tgeoghegan) [inline facilitator/src/sample.rs:134]: Rather than list other magic numbers, this comment should concisely explain why we're using a constant. "running an instance of libprio::Server" is kind of vague, too.

File: facilitator/src/bin/facilitator.rs
Comment (tgeoghegan) [inline facilitator/src/bin/facilitator.rs:460]: The environment variable names should match the parameter names.

Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs:460]: Oh whoops

File: facilitator/src/bin/facilitator.rs
Comment (tgeoghegan) [inline facilitator/src/bin/facilitator.rs:2]: These changes to `facilitator` make sense, but they will break existing paired test envs until the corresponding Terraform changes land

**Pull request closed without merging**
